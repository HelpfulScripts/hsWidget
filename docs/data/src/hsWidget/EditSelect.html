
                <html><style>body { overflow:hidden;}
            h1 { font-family: Arial, sans-serif; font-size: 24px; color: #44a; }
            p { margin:0; padding:0; }
            br  { margin:0; padding:0; }
            .line { 
                white-space: nowrap;
                height:16px; 
            }
            .line>span { 
                display:inline-block; background-color:#eef; height:100%; 
                margin: 0 5px 0 0; padding-right: 5px; color:#999;   
            }
            comment {color: #080;} module {color: #804;} 
            quote {color: #008;} comment>quote {color: #080;}
            .listing { margin: 10px; border: 1px solid #ccc; 
                    font-family: SFMono-Regular, Consolas, 'Liberation Mono', Menlo, Courier, monospace;
                    font-size: 14px;  
                    overflow:scroll;
                    height:90%;
            }
            code { padding: 5px 0;}</style></html><body>
                <h1>EditSelect.ts</h1>
                <div class='listing'><code><p id=1 class="line"><span>&nbsp;&nbsp;&nbsp;1</span><comment>/**</comment></p>
<p id=2 class="line"><span>&nbsp;&nbsp;&nbsp;2</span><comment>&nbsp;*&nbsp;#&nbsp;EditSelect</comment></p>
<p id=3 class="line"><span>&nbsp;&nbsp;&nbsp;3</span><comment>&nbsp;*&nbsp;Provides&nbsp;a&nbsp;text&nbsp;label&nbsp;with&nbsp;pulldown&nbsp;select&nbsp;capability</comment></p>
<p id=4 class="line"><span>&nbsp;&nbsp;&nbsp;4</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=5 class="line"><span>&nbsp;&nbsp;&nbsp;5</span><comment>&nbsp;*&nbsp;###&nbsp;Attributes:</comment></p>
<p id=6 class="line"><span>&nbsp;&nbsp;&nbsp;6</span><comment>&nbsp;*&nbsp;-&nbsp;from:&nbsp;`string[]`&nbsp;the&nbsp;array&nbsp;to&nbsp;select&nbsp;from</comment></p>
<p id=7 class="line"><span>&nbsp;&nbsp;&nbsp;7</span><comment>&nbsp;*&nbsp;-&nbsp;selected:&nbsp;`string`&nbsp;the&nbsp;initial&nbsp;string&nbsp;to&nbsp;show&nbsp;in&nbsp;the&nbsp;text&nbsp;label</comment></p>
<p id=8 class="line"><span>&nbsp;&nbsp;&nbsp;8</span><comment>&nbsp;*&nbsp;-&nbsp;css:&nbsp;the&nbsp;css&nbsp;selector&nbsp;to&nbsp;add&nbsp;to&nbsp;the&nbsp;field</comment></p>
<p id=9 class="line"><span>&nbsp;&nbsp;&nbsp;9</span><comment>&nbsp;*&nbsp;-&nbsp;update:&nbsp;callback&nbsp;function&nbsp;called&nbsp;when&nbsp;the&nbsp;input&nbsp;field&nbsp;looses&nbsp;focus</comment></p>
<p id=10 class="line"><span>&nbsp;&nbsp;10</span><comment>&nbsp;*&nbsp;-&nbsp;**popup**?:&nbsp;`string`:&nbsp;optional&nbsp;message&nbsp;to&nbsp;show&nbsp;in&nbsp;a&nbsp;{@link&nbsp;Popup&nbsp;popup&nbsp;box}&nbsp;upon&nbsp;mouseover.</comment></p>
<p id=11 class="line"><span>&nbsp;&nbsp;11</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=12 class="line"><span>&nbsp;&nbsp;12</span><comment>&nbsp;*&nbsp;###&nbsp;Example</comment></p>
<p id=13 class="line"><span>&nbsp;&nbsp;13</span><comment>&nbsp;*&nbsp;&lt;example&gt;</comment></p>
<p id=14 class="line"><span>&nbsp;&nbsp;14</span><comment>&nbsp;*&nbsp;&lt;file&nbsp;name=<quote>'script.js'</quote>&gt;&nbsp;</comment></p>
<p id=15 class="line"><span>&nbsp;&nbsp;15</span><comment>&nbsp;*&nbsp;const&nbsp;choices&nbsp;=&nbsp;[<quote>'pizza'</quote>,&nbsp;<quote>'chicken'</quote>,&nbsp;<quote>'fries'</quote>];</comment></p>
<p id=16 class="line"><span>&nbsp;&nbsp;16</span><comment>&nbsp;*&nbsp;let&nbsp;selected&nbsp;=&nbsp;<quote>'none'</quote>;</comment></p>
<p id=17 class="line"><span>&nbsp;&nbsp;17</span><comment>&nbsp;*&nbsp;m.mount(root,&nbsp;{view:&nbsp;()&nbsp;=&gt;&nbsp;m(<quote>'div'</quote>,&nbsp;[</comment></p>
<p id=18 class="line"><span>&nbsp;&nbsp;18</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'h3'</quote>,&nbsp;<quote>'Edit&nbsp;Select&nbsp;Example'</quote>),</comment></p>
<p id=19 class="line"><span>&nbsp;&nbsp;19</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'div'</quote>,&nbsp;[</comment></p>
<p id=20 class="line"><span>&nbsp;&nbsp;20</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'span'</quote>,&nbsp;`current&nbsp;content:&nbsp;<quote>'${selected}'</quote>`),</comment></p>
<p id=21 class="line"><span>&nbsp;&nbsp;21</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(hsWidget.EditSelect,&nbsp;{</comment></p>
<p id=22 class="line"><span>&nbsp;&nbsp;22</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;class:&nbsp;<quote>'mySelect'</quote>,</comment></p>
<p id=23 class="line"><span>&nbsp;&nbsp;23</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;initial:&nbsp;selected,</comment></p>
<p id=24 class="line"><span>&nbsp;&nbsp;24</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;update:&nbsp;newValue&nbsp;=&gt;&nbsp;selected&nbsp;=&nbsp;newValue</comment></p>
<p id=25 class="line"><span>&nbsp;&nbsp;25</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;},&nbsp;choices)</comment></p>
<p id=26 class="line"><span>&nbsp;&nbsp;26</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;])</comment></p>
<p id=27 class="line"><span>&nbsp;&nbsp;27</span><comment>&nbsp;*&nbsp;])});</comment></p>
<p id=28 class="line"><span>&nbsp;&nbsp;28</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=29 class="line"><span>&nbsp;&nbsp;29</span><comment>&nbsp;*&nbsp;&lt;/file&gt;</comment></p>
<p id=30 class="line"><span>&nbsp;&nbsp;30</span><comment>&nbsp;*&nbsp;&lt;file&nbsp;name=<quote>'style.css'</quote>&gt;</comment></p>
<p id=31 class="line"><span>&nbsp;&nbsp;31</span><comment>&nbsp;*&nbsp;.mySelect&nbsp;{&nbsp;&nbsp;</comment></p>
<p id=32 class="line"><span>&nbsp;&nbsp;32</span><comment>&nbsp;*&nbsp;}</comment></p>
<p id=33 class="line"><span>&nbsp;&nbsp;33</span><comment>&nbsp;*&nbsp;&lt;/file&gt;</comment></p>
<p id=34 class="line"><span>&nbsp;&nbsp;34</span><comment>&nbsp;*&nbsp;&lt;/example&gt;</comment></p>
<p id=35 class="line"><span>&nbsp;&nbsp;35</span><comment>&nbsp;*/</comment></p>
<p id=36 class="line"><span>&nbsp;&nbsp;36</span></p>
<p id=37 class="line"><span>&nbsp;&nbsp;37</span><comment>/**&nbsp;*/</comment></p>
<p id=38 class="line"><span>&nbsp;&nbsp;38</span>import&nbsp;m&nbsp;from&nbsp;<quote>"mithril"</quote>;</p>
<p id=39 class="line"><span>&nbsp;&nbsp;39</span>import&nbsp;{&nbsp;Log&nbsp;}&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;<quote>'hsutil'</quote>;&nbsp;&nbsp;const&nbsp;log&nbsp;=&nbsp;new&nbsp;Log(<quote>'EditSelect'</quote>);</p>
<p id=40 class="line"><span>&nbsp;&nbsp;40</span>import&nbsp;{&nbsp;Popup&nbsp;}&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;<quote>'./Popup'</quote>;</p>
<p id=41 class="line"><span>&nbsp;&nbsp;41</span>import&nbsp;{&nbsp;Widget,&nbsp;WidgetAttrs,&nbsp;ViewResult&nbsp;}&nbsp;&nbsp;&nbsp;from&nbsp;<quote>"./Widget"</quote>;</p>
<p id=42 class="line"><span>&nbsp;&nbsp;42</span>import&nbsp;{&nbsp;Vnode&nbsp;}&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;<quote>"./Widget"</quote>;</p>
<p id=43 class="line"><span>&nbsp;&nbsp;43</span>&nbsp;&nbsp;</p>
<p id=44 class="line"><span>&nbsp;&nbsp;44</span></p>
<p id=45 class="line"><span>&nbsp;&nbsp;45</span></p>
<p id=46 class="line"><span>&nbsp;&nbsp;46</span>export&nbsp;interface&nbsp;EditSelectAttrs&nbsp;extends&nbsp;WidgetAttrs&nbsp;{</p>
<p id=47 class="line"><span>&nbsp;&nbsp;47</span>&nbsp;&nbsp;&nbsp;&nbsp;popup?:&nbsp;string;</p>
<p id=48 class="line"><span>&nbsp;&nbsp;48</span>&nbsp;&nbsp;&nbsp;&nbsp;update:&nbsp;(r:string)&nbsp;=&gt;&nbsp;void;</p>
<p id=49 class="line"><span>&nbsp;&nbsp;49</span>&nbsp;&nbsp;&nbsp;&nbsp;initial:&nbsp;string;</p>
<p id=50 class="line"><span>&nbsp;&nbsp;50</span>}</p>
<p id=51 class="line"><span>&nbsp;&nbsp;51</span></p>
<p id=52 class="line"><span>&nbsp;&nbsp;52</span>export&nbsp;class&nbsp;EditSelect&nbsp;extends&nbsp;Widget&nbsp;{</p>
<p id=53 class="line"><span>&nbsp;&nbsp;53</span>&nbsp;&nbsp;&nbsp;&nbsp;selectable:boolean;</p>
<p id=54 class="line"><span>&nbsp;&nbsp;54</span>&nbsp;&nbsp;&nbsp;&nbsp;select:&nbsp;(e:Event)=&gt;void</p>
<p id=55 class="line"><span>&nbsp;&nbsp;55</span>&nbsp;&nbsp;&nbsp;&nbsp;oninit(node:Vnode&lt;EditSelectAttrs,&nbsp;this&gt;)&nbsp;{</p>
<p id=56 class="line"><span>&nbsp;&nbsp;56</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;node.state.selectable&nbsp;=&nbsp;false;</p>
<p id=57 class="line"><span>&nbsp;&nbsp;57</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;node.state.select&nbsp;=&nbsp;(e:Event)&nbsp;=&gt;&nbsp;{&nbsp;</p>
<p id=58 class="line"><span>&nbsp;&nbsp;58</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;node.attrs.update((&lt;HTMLButtonElement&gt;e.currentTarget).value);</p>
<p id=59 class="line"><span>&nbsp;&nbsp;59</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;node.state.selectable&nbsp;=&nbsp;false;</p>
<p id=60 class="line"><span>&nbsp;&nbsp;60</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=61 class="line"><span>&nbsp;&nbsp;61</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<comment>//&nbsp;node.state.makeSelectable&nbsp;=&nbsp;()&nbsp;=&gt;&nbsp;{</comment></p>
<p id=62 class="line"><span>&nbsp;&nbsp;62</span><comment></comment>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<comment>//&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;node.state.selectable&nbsp;=&nbsp;true;</comment></p>
<p id=63 class="line"><span>&nbsp;&nbsp;63</span><comment></comment>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<comment>//&nbsp;}</comment></p>
<p id=64 class="line"><span>&nbsp;&nbsp;64</span><comment></comment>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=65 class="line"><span>&nbsp;&nbsp;65</span>&nbsp;&nbsp;&nbsp;&nbsp;view(node:Vnode&lt;EditSelectAttrs,&nbsp;this&gt;):ViewResult&nbsp;{</p>
<p id=66 class="line"><span>&nbsp;&nbsp;66</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;m(`select.hsedit_select`,&nbsp;</p>
<p id=67 class="line"><span>&nbsp;&nbsp;67</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Popup.arm(node.attrs.popup,&nbsp;this.attrs(node.attrs,&nbsp;&lt;any&gt;{&nbsp;onchange:node.state.select})),</p>
<p id=68 class="line"><span>&nbsp;&nbsp;68</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(&lt;string[]&gt;node.children).map((o:string)&nbsp;=&gt;&nbsp;node.attrs.initial===o?</p>
<p id=69 class="line"><span>&nbsp;&nbsp;69</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'option.hsedit_select_option.selected'</quote>,&nbsp;{&nbsp;value:&nbsp;o,&nbsp;selected:true&nbsp;},&nbsp;o)&nbsp;:</p>
<p id=70 class="line"><span>&nbsp;&nbsp;70</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'option.hsedit_select_option'</quote>,&nbsp;{&nbsp;value:&nbsp;o&nbsp;},&nbsp;o)));</p>
<p id=71 class="line"><span>&nbsp;&nbsp;71</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=72 class="line"><span>&nbsp;&nbsp;72</span>}</p></code></div>
                </body>
            