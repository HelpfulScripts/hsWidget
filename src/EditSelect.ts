/**
 * # EditSelect
 * Provides a text label with pulldown select capability
 * 
 * ### Attributes:
 * - from: `string[]` the array to select from
 * - selected: `string` the initial string to show in the text label
 * - css: the css selector to add to the field
 * - update: callback function called when the input field looses focus
 * - **popup**?: `string`: optional message to show in a {@link Popup popup box} upon mouseover.
 * 
 * ### Example
 * <example>
 * <file name='script.js'> 
 * const choices = ['pizza', 'chicken', 'fries'];
 * let selected = 'none';
 * m.mount(root, {view: () => m('div', [
 *    m('h3', 'Edit Select Example'),
 *    m('div', [
 *      m('span', `current content: '${selected}'`),
 *      m(hsWidget.EditSelect, {
 *          class: 'mySelect',
 *          initial: selected,
 *          update: newValue => selected = newValue
 *      }, choices)
 *    ])
 * ])});
 * 
 * </file>
 * <file name='style.css'>
 * .mySelect {  
 * }
 * </file>
 * </example>
 */

/** */
import m from "mithril";
import { Log }      from 'hsutil';  const log = new Log('EditSelect');
import { Popup }    from './Popup';
import { Widget, WidgetAttrs, ViewResult }   from "./Widget";
import { Vnode }    from "./Widget";
  


export interface EditSelectAttrs extends WidgetAttrs {
    popup?: string;
    update: (r:string) => void;
    initial?: string;
}

export class EditSelect extends Widget {
    selectable:boolean;
    select: (e:Event)=>void
    selected:string;
    oninit(node:Vnode<EditSelectAttrs, this>) {
        node.state.selectable = false;
        const i = (<string[]>node.children).indexOf(node.attrs.initial);
        node.state.selected = (<string[]>node.children)[i>=0? i : 0];
        node.state.select = (e:Event) => { 
            const selection = (<HTMLButtonElement>e.currentTarget).value;
            node.attrs.update(selection);
            node.state.selected = selection;
            node.state.selectable = false;
        }
    }
    view(node:Vnode<EditSelectAttrs, this>):ViewResult {
        return m(`select.hsedit_select`, 
            Popup.arm(node.attrs.popup, this.attrs(node.attrs, <any>{ onchange:node.state.select})),
            (<string[]>node.children).map((o:string) => node.state.selected===o?
                m('option.hsedit_select_option.selected', { value: o, selected:true }, o) :
                m('option.hsedit_select_option', { value: o }, o)));
    }
}