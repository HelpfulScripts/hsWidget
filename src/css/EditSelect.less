

.hsedit_select {
    border:0;
    background-color: rgba(0,0,0,0.1);
    font-family: inherit;
    font-size: inherit;
}

// .hsedit_select_option {
//     &.selected {

//     }
// }