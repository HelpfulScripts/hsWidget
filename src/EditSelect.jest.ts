import { m }      from 'hslayout';
import { EditSelect }    from './EditSelect';
const root = window.document.createElement("div");

const choices = ['pizza', 'chicken', 'fries'];
let selected = 'none';

describe('EditSelect', () => {
    beforeAll(()=>{
        m.mount(root, { view: () => m(EditSelect, {
            css: '.mySelect',
            from: choices,
            selected: selected,
            update: (newValue:string) => selected = newValue
        })});
    });

    it('should match snapshot', () => {
        expect(root).toMatchSnapshot();
    });
});