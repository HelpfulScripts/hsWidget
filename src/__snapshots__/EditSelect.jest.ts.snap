// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSelect with initial should match snapshot 1`] = `
<div>
  <select
    class="hsedit_select  mySelect"
  >
    <option
      class="hsedit_select_option"
      value="pizza"
    >
      pizza
    </option>
    <option
      class="hsedit_select_option"
      value="chicken"
    >
      chicken
    </option>
    <option
      class="hsedit_select_option selected"
      value="fries"
    >
      fries
    </option>
  </select>
</div>
`;

exports[`EditSelect without initial should match snapshot 1`] = `
<div>
  <select
    class="hsedit_select  mySelect"
  >
    <option
      class="hsedit_select_option selected"
      value="pizza"
    >
      pizza
    </option>
    <option
      class="hsedit_select_option"
      value="chicken"
    >
      chicken
    </option>
    <option
      class="hsedit_select_option"
      value="fries"
    >
      fries
    </option>
  </select>
</div>
`;
